src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/CacheableWhitelist.java

```diff
@@ -0,0 +1,10 @@
+package org.jenkinsci.plugins.scriptsecurity.sandbox.whitelists;
+
+/**
+ * Marker interface indicating results of permission checks do not change unless the list of entries is modified
+ *  and depend ONLY on the Field/Method/Constructor actually being invoked, irrespective of classloader.
+ *  That means you must be able to represent the WhiteList entry with a String signature of className + method/field signature.
```

Comment:
Setting aside `BlanketWhitelist`, which is not actually used in production and so need not actually be marked, I think what you really mean is that this should be applied only to `EnumeratingWhitelist`s, or to `ProxyWhitelist`s which only ever set their delegate once, and to an `EnumeratingWhitelist`. And these latter could actually be fixed to use the otherwise unusual field-scoped `@Extension`:

```java
public class JenkinsWhitelist {
    @Extension
    public static final Whitelist INSTANCE;
    static {
        try {
            INSTANCE = StaticWhitelist.from(JenkinsWhitelist.class.getResource("jenkins-whitelist"));
        } catch (IOException x) {
            throw new ExceptionInInitializerError(x);
        }
    }
```

Furthermore, change `EnumeratingWhitelist` to cache its signatures on demand, rather than repeatedly calling the methods. Since its check methods are `final`, we can then be assured that it is really safe to cache calls to it. Then delete `CacheableWhitelist` (so there is no chance of someone misusing it, as you accidentally did earlier in this PR! too easy to miss such subtleties) and simply check for `EnumeratingWhitelist` from `ProxyWhitelist`. Which we already do (including sorting the `EnumeratingWhitelist`s at the top of the list, since it is really a set)—so perhaps most of the rest of this PR could be discarded as well, assuming the only real problem at runtime was that we had some things like `JenkinsWhitelist` that were opaque to `ProxyWhitelist`’s optimizations for a fairly silly reason.